mlir/lib/IR/PatternMatch.cpp

```diff
@@ -127,6 +127,35 @@ void RewriterBase::replaceAllOpUsesWith(Operation *from, Operation *to) {
   replaceAllUsesWith(from->getResults(), to->getResults());
 }
 
+ReplaceOpAction::ReplaceOpAction(ArrayRef<IRUnit> irUnits,
+                                 ValueRange replacement)
+    : Base(irUnits), replacement(replacement) {
+  assert(irUnits.size() == 1);
+  assert(irUnits[0]);
+  assert(isa<Operation *>(irUnits[0]));
+}
+
+void ReplaceOpAction::print(raw_ostream &os) const {
+  OpPrintingFlags flags;
+  flags.elideLargeElementsAttrs(10);
+  os << "`" << tag << "` replacing operation `";
+  getOp()->print(os, flags);
+  os << "` by ";
+  bool first = true;
+  for (auto r : replacement) {
+    if (!first)
+      os << ", ";
+    os << "`";
+    r.print(os, flags);
+    os << "`";
+    first = false;
+  }
```

Comment:
nit: This could have been written with `llvm::interleave`, but does not make a big difference